### DISCUSSION TITLE
Best way to handle dual core chips?

### DISCUSSION THREAD

**Discussion by radiohound** [stm32duino/Arduino_Core_STM32#1593] [Variants]
I now have one of the dual core H747xi devices working on its CM7 core. I'm a bit perplexed how best to handle the dual core aspects. It seems like the variant file (H742X(G-I)H_H743X(G-I)H_H745X(G-I)H_H747X(G-I)H_H750XBH_H753XIH_H755XIH_H757XIH) needs to be duplicated. One for a core_cm4 and one for core_cm7, and have two board definitions for them in the board text file. That way, you would select for each board (depending on which core you are using. This would be somewhat similar to how the user selects which core they are working with the Portenta, and the Sony Spresense devices.

There are some complications that are hard (for me) to figure out with dual cores. I believe there also needs to be two different include files stm32h747xx.h . A line added into these files ... define CORE_CM4 and the other one with define CORE_CM7. I am a newbie, and not really sure how that should be implemented inside your current file structure. I don't understand how these includes are pulled when you compile with the arduino ide, and stm32duino. 

Any help you can offer will be greatly appreciated. Thank you.

**Comment by ABOSTM** [comment]
Hi @radiohound, 
This Arduino core, stm32duino/Arduino_Core_STM32, does not support dual core at all.
With dual core, comes some tricky system issues to worry about, like:
* Clock tree is shared between both cores. so usually, one core (CM7 for example) configure the clocks for both cores.
But this implies that the other core (CM4) wait for end of clock initialisation before to go further.
This implies some hardware synchronisation between both cores (thanks to Hardware semaphore HSEM, ...).
*  Other peripherals are shared between cores, and every write to a common register should be protected with HSEM too.

* Boot sequence: either both core boots at the same time (and a synchro is needed), either one core boots first and it is in charge of booting the second one ...
* memory split (Flash, Ram), eventually memory sharing
* etc ...

To be frank, if as you said you are a newbie, I would suggest you to start by using H747xi devices as a single core (CM7) for a while and later, once you feel more confortable, jump to dual core. Complex system issues, like described above, which are not properly handled, could become a nightmare to debug on dualcore.
Arduino could be fine if you use MCU as single core CM7, but I would suggest you to not use Arduino for this chip if you intend to use dual core .
If you really wants to deal with dual core, there are some alternatives I can suggest you:
* Official STM32cube drivers
* Mbed
* Zephyr
